# SDO Machine Learning (ML) Dataset

In this notebook we demonstrate the process for interacting with a small sample of the SDO ML dataset.

*Meng Jin, Mark Cheung, & Paul Wright*

---

## Introduction

### SDO Overview

Since its launch in 2010, NASA’s Solar Dynamics Observatory (SDO; Pesnell et al. 2012) has continuously monitored Sun's activity, delivering a wealth of valuable scientific data for heliophysics researchers with the use of three instruments:

1. The Atmospheric Imaging Assembly (AIA; Lemen et al. 2012), which captures 4096 x 4096 resolution images (with 0.6 arcsecond pixel size) of the full Sun in two ultraviolet (centered at 1600, and 1700 Å), seven extreme ultraviolet (EUV; centered at 94, 131, 171, 193, 211, 304, and 335 Å), and one visible (centered at 4500 Å) wavelength band.

2. The Helioseismic and Magnetic Imager (HMI; Schou et al. 2012) captures visible wavelength filtergrams of the full Sun at 4096 x 4096 resolution (a pixel size of 0.5 arcsecond), which are then processed into a number of data products, including photospheric Dopplergrams, line-of-sight magnetograms, and vector magnetograms (Hoeksema et al. 2014).

3. The EUV Variability Experiment (EVE; Woods et al. 2012) monitors the solar EUV spectral irradiance from 1 to 1050 Å. This is done by utilizing multiple EUV Grating Spectrographs (MEGS) that disperse EUV light from the full disk of the Sun and its corona onto a 1024 x 2048 charge coupled device (CCD).

### The SDO ML Dataset
The SDO ML Dataset (covering 2010 - 2018) was originally published as Galvez et al (2019), and is hosted on the Stanford Digital Repository in Numpy's compressed array format (.npz). 

In this notebook, we present an update to the work outlined in [Galvez et al (2019)](https://ui.adsabs.harvard.edu/abs/2019ApJS..242....7G/abstract). Specifically, we will concentrate on the SDO/AIA data, as this has been updated to account for a change in calibration after 20??. In addtion to the change in calibration, this updated format includes

1. FITS header/keyword information (such as observation time, and exposure time).
2. processes for continually updating the data until the present day.

### Who is the SDO ML Dataset for?

The sheer volume of structured scientific data recorded by SDO (over 18 PB, and counting) is ideal for a range machine learning tasks (from time-series, to computer vision), as well as more traditional approaches.

While the level 1 data are easily accessible, pre-processing these data for scientific analysis often requires  specialized heliophysics (and instrument-specific) knowledge. This may act as an unnecessary hurdle for non-heliophysics machine learning researchers who may wish to experiment with datasets from the physical sciences, but are unaware of domain-specific nuances (e.g., that images must be spatially and temporally adjusted).

This notebook demonstrates the process for interacting with a subset of the curated SDO (AIA) dataset, that is mission-ready for machine-learning applications. **Our aim is to supply this standardized dataset for heliophysicists who wish to use machine learning in their own research, as well as machine-learning researchers who wish to develop models specialized for the physical sciences.**

---

## Table of Contents

The notebook is set out as follows:

1. Setting up the notebook
2. Reading and loading the data <br>
2a. Selecting images based on header information <br>
    2b. Selecting a subset of the data based on index <br>
    2c. Downsampling the data <br>
    2d. Downsampling the data <br>
3. Generating a SunPy map
4. Animating AIA data

## 1. Setting up the notebook

In [3]:
!pip install zarr
!pip install sunpy[all]
!pip install gcsfs

In [4]:
%matplotlib inline

import gcsfs
import zarr
import dask.array as da
import numpy as np
import sunpy.map 
import matplotlib.pyplot as plt
import pandas as pd

from sunpy.visualization import axis_labels_from_ctype, wcsaxes_compat

from matplotlib import animation
from IPython.display import HTML

In [5]:
#not sure if this actually works
from google.cloud import storage
from google.colab import auth

auth.authenticate_user()

ModuleNotFoundError: No module named 'google.cloud'

In [6]:
# this is currently fine for interacting with notebooks from within the project
gcs = gcsfs.GCSFileSystem(access='read_only')
loc = 'gs://fdl-sdoml-v2/sdomlv2_small.zarr'
store = gcsfs.GCSMap(loc, gcs=gcs, check=False)

## 2. Reading and loading the data

The SDO ML dataset is stored in the Zarr format, a format for the storage of chunked, compressed, N-dimensional arrays with Numpy dtype. For an in-depth overview, see https://zarr.readthedocs.io/en/stable/tutorial.html.

In [7]:
#first, we create a group with the store data located on GCP.
root = zarr.group(store)

As shown in the tree, the heirachy consists of groups, each shown with their respective shape, and data type. In this example, we will primarily look at the 171 Å channel from 2010. This consists of 6135 512x512 images, stored as float32, and can be accessed as follows:

In [8]:
data = root['2010']['171A']

We could have alternatively accessed the 2010 data as follows 

```
loc = 'fdl-sdoml-v2/sdomlv2_small.zarr/2010'
```

which becomes increasingly useful in the full dataset (where the heirachy contains years 2010 - present).

**Loading with Dask**

We can then load this data into an array using dask.

In [9]:
da.from_array(data)

dask.array<array, shape=(6135, 512, 512), dtype=float32, chunksize=(120, 512, 512), chunktype=numpy.ndarray>

As shown above, the data has the shape (6135, 512, 512), and is split into 52 chunks of (120, 512, 512), each of 125.83 MB; this is further visualised on the right. The data is now in a form to be manipulated like a Numpy array. 


Depending on the use-case, we may wish to extract a subset of this data in various ways. In the following sections we step through a number of potential operations that we may wish to make with the data.

### 2a. Selecting images based on header information

We can extract the exposure (and observation) time from the data attributes (the header information), and downsample our data based upon that information.

In [10]:
exptime=np.array(data.attrs['EXPTIME'])
t_obs=np.array(data.attrs['T_OBS'])

In [11]:
#TODO: Visualise exposure time

In [12]:
# select indices where the exposure time is less than 2 seconds
index=np.where(exptime<2.0)
selected_images=da.from_array(data)[index[0],:,:]

In [13]:
selected_images

dask.array<getitem, shape=(4005, 512, 512), dtype=float32, chunksize=(120, 512, 512), chunktype=numpy.ndarray>

### 2b. Selecting images based on indices

While the data is not currently ordered by observation time, we can simple index the array to extract a number of observations

In [14]:
# select a subset of the data based on index
selected_images = da.from_array(data)[100:200,:,:]
selected_header = {keys:values[100:200] for keys,values in data.attrs.items()}

# and even compute the mean of the selected images
print(f'mean of selected images: {da.mean(selected_images).compute():.2f}')

mean of selected images: 107.05


In [15]:
selected_images

dask.array<getitem, shape=(100, 512, 512), dtype=float32, chunksize=(80, 512, 512), chunktype=numpy.ndarray>

### 2c. Downsampling the data (resolution)

In [16]:
# downsample the data to 256 by sampling every other pixel
sub_index = np.arange(0,512,2)

sub_image=da.from_array(data)[0:600,sub_index,:]
sub_image=sub_image[:,:,sub_index]
# is this becuase dask doesn't support nd fancy indexing?

print(f'mean of selected images: {da.mean(sub_image).compute():.2f}')

mean of selected images: 113.65


In [17]:
sub_image

dask.array<getitem, shape=(600, 256, 256), dtype=float32, chunksize=(120, 256, 256), chunktype=numpy.ndarray>

### 2d. Downsampling the data (temporally)

In [ ]:
df_time=pd.DataFrame(t_obs,index=np.arange(np.shape(t_obs)[0]),columns=['Time'])
df_time['Time'] = pd.to_datetime(df_time['Time'])

# select times at a frequency of 12 minutes 
selected_times = pd.date_range(start='2010-08-28 00:00:00', end='2010-08-28 23:59:59', freq='12T',tz='UTC')

In [ ]:
selected_index = []
for i in selected_times:
    selected_index.append(np.argmin(abs(df_time['Time']-i)))
# may be a more efficient way of doing this
time_index = list(filter(lambda x : x > 0, selected_index))
da.from_array(data)[time_index,:,:]

## 3. Generating a SunPy Map

SunPy is an open-source Python library for Solar Physics data analysis and visualization. 

In this section we will demonstrate how SunPy’s [Map](https://docs.sunpy.org/en/stable/api/sunpy.map.Map.html#sunpy.map.Map) with the Zarr-formatted data. We demonstrate this for a single index.

In [ ]:
#specify an image index
img_index = time_index[-1]

In [ ]:
#select the respective image, and header required for sunpy.map.Map()
selected_image = da.from_array(data)[img_index,:,:]
selected_headr = {keys:values[img_index] for keys,values in data.attrs.items()}

In [ ]:
my_map = sunpy.map.Map((np.array(selected_image), selected_headr))

In [ ]:
plt.figure(figsize=(7,7))
ax = plt.subplot(projection=my_map)
my_map.plot()

As this is then a SunPy Map object, we can manipulate it as such.

In [ ]:
my_map

For more information about SunPy, and Map,

## 4. Generating an AIA animation

In [ ]:
maps = []

for i in time_index:
    selected_image = da.from_array(data)[i,:,:]
    selected_headr = {keys:values[i] for keys,values in data.attrs.items()}
    maps.append(sunpy.map.Map((np.array(selected_image), selected_headr)))
    
sq = sunpy.map.Map(maps, sequence=True, sortby="date")   

In [ ]:
def animate_sequence(sequence):
    """
    An animation plotting routine to animate each element in MapSequence
    
    Parameters
    ----------
    sequence : `sunpy.map.mapsequence.MapSequence`
        a set of sunpy maps to animate
        
    Returns
    -------
    `matplotlib.animation.FuncAnimation`
        A FuncAnimation instance.
    """

    #fig, ax = plt.subplots(figsize=(7,7), 
    #                       subplot_kw=dict(projection=sequence[0].wcs)
    #                      ) 

    ax = wcsaxes_compat.gca_wcs(sequence[0].wcs)
    fig = ax.get_figure()
    fig.figsize = (7,7)

    plot_obj = sequence[0].plot()
    ax.set_title(sequence[0].latex_name)
    
    cbar = plt.colorbar()

    def update_fig(i):
        ax.set_title(sequence[i].latex_name)
    
        # set the data to that of the image being plot
        # Each image should be scaled as the first.
        plot_obj.set_data(sequence[i].data)
    
        # reset the WCS to that of the image being plot
        plot_obj.axes.reset_wcs(sequence[i].wcs)
        
        wcsaxes_compat.default_wcs_grid(ax)
        ax.set_xlabel(axis_labels_from_ctype(sequence[i].coordinate_system[0],
                                               sequence[i].spatial_units[0]))
        ax.set_ylabel(axis_labels_from_ctype(sequence[i].coordinate_system[1],
                                               sequence[i].spatial_units[1]))

        #basic test:
        #expect the wcs of the plot_obj to follow the sequence
        #print(sequence[i].wcs == plot_obj.axes.wcs, sequence[0].wcs == sequence[i].wcs)
        return plot_obj,

    anim = animation.FuncAnimation(fig, update_fig, init_func=None,
                                   frames=len(sequence), interval=60, repeat=True
                                  )
    plt.close(fig)
    
    return anim


In [ ]:
anim = animate_sequence(sq)
HTML(anim.to_jshtml())

----

This project was conducted during the 2018 NASA Frontier Development Lab (FDL) program, a public/private partnership between NASA and SETI and industry partners including NVIDIA Corporation, Lockheed Martin, and IBM. The authors thank IBM (especially Naeem Altaf) for generously providing computing resources on the IBM Cloud.